util/storage-writer/src/csv_storage_writer.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::io;
+use std::fs;
+use std::fs::File;
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+use super::{Database,StorageWriter,StorageWriterConfig};
+use csv::Writer;
+use ethereum_types::{Address, H256};
+
+/// Implementation of `StorageWriter` trait for writing to CSV file.
+#[derive(Clone)]
+pub struct CsvStorageWriter {
+    /// Path to CSV
+    path: PathBuf,
+    /// Contracts for which to write storage diffs
+    watched_contracts: Vec<Address>,
+    /// File writing connection to CSV
+    writer: Arc<Mutex<Writer<File>>>
+}
+
+impl CsvStorageWriter {
+    pub fn new(config: StorageWriterConfig) -> CsvStorageWriter {
+        let path = config.path;
+        let file = path.join("storage_diffs.csv");
+
+        // TODO: handle error differently on file creation
+        let file = fs::OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .append(true)
+            .open(file)
+            .expect("Error creating csv file.");
+
+        let wtr = csv::Writer::from_writer(file);
+
+        CsvStorageWriter {
+            path: path,
+            watched_contracts: config.watched_contracts,
+            writer: Arc::new(Mutex::new(wtr))
+        }
+    }
+
+    fn write_storage_node(&mut self, contract: Address, block_hash: H256, block_number: u64, key: H256, value: H256) -> io::Result<()> {
+        let mut wtr = self.writer.lock().unwrap();
+        wtr.write_record(&[format!("{:x}", contract), format!("{:x}", block_hash), format!("{}", block_number), format!("{:x}", key), format!("{:x}", value)])?;
+        wtr.flush()?;
+        Ok(())
+    }
+}
+
+impl StorageWriter for CsvStorageWriter {
+    fn boxed_clone(&self) -> Box<StorageWriter> {
+        let config = StorageWriterConfig {
+            database: Database::Csv,
+            enabled: true,
+            path: self.path.clone(),
+            watched_contracts: self.watched_contracts.to_vec(),
+        };
+        Box::new(CsvStorageWriter::new(config))
+    }
+
+    fn enabled(&self) -> bool {
+        true
```

Comment:
idk the exact syntax but should this be read from the `config` instance variable? `self.enabled`

Reply:
yeah good call - the config variable was added later and I totally forgot we could reuse

Reply:
actually on a second look it seems like the config is not actually an instance variable on the storage writer. sorta torn between leaving it hardcoded vs making config an instance variable and nesting all the lookups onto that 🤔

Reply:
oooo i see what youre saying. what you think about extracting just the `enabled` field from the config and set that as an instance variable instead of making the whole config an instance variable.

Reply:
if it's alright with you I think I may postpone making that change. My thinking is that one of these should only ever be initialized if it's enabled (otherwise it should be the noop version being used), and the rationale for having it on the config at all is just so that we can make decisions about whether or not to create a directory for the diffs from a different namespace (that doesn't have access to the writer itself). thinking we could always make it more directly injected if a use case arises, but for now I'd rather not make the result of `enabled() == true` depend on values being passed from far reaches of the codebase

Reply:
👍 sounds good